mapo/agents/mapo/mapo_policy.py

```diff
@@ -0,0 +1,143 @@
+        )
+    )
+
+    # DPG loss
+    actor_loss = -tf.reduce_mean(policy.q_values(policy.actions))
+
+    return actor_loss + critic_loss
+
+
+def check_action_space(policy, obs_space, action_space, config):
+    # pylint: disable=unused-argument
+    """Check if the action space is suited to DPG."""
+    if not isinstance(action_space, Box):
+        raise UnsupportedSpaceException(
+            "Action space {} is not supported for MAPO.".format(action_space)
+        )
+    if len(action_space.shape) > 1:
+        raise UnsupportedSpaceException(
+            "Action space has multiple dimensions {}.".format(action_space.shape)
+            + "Consider reshaping this into a single dimension, using a Tuple action"
+            "space, or the multi-agent API."
+        )
+
+
+def build_continuous_q_function(input_dict, obs_space, action_space, config):
+    """Construct the state preprocessor and critic layers."""
+    if config["use_state_preprocessor"]:
+        state_preprocessor = ModelCatalog.get_model(
+            input_dict, obs_space, action_space, 1, config["model"]
+        )
+        processed_obs = state_preprocessor.last_layer
+    else:
+        state_preprocessor = None
+        processed_obs = input_dict[SampleBatch.CUR_OBS]
+
+    activation = config["critic_hidden_activation"]
+    hiddens = [
+        keras.layers.Dense(units=hidden, activation=activation)
+        for hidden in config["critic_hiddens"]
+    ]
+    layers = hiddens + [keras.layers.Dense(units=1, activation=None)]
+    model = keras.Sequential(layers=layers)
+
+    def evaluate_actions(actions):
+        return model(tf.concat([processed_obs, actions], axis=1))
+
+    return evaluate_actions, state_preprocessor
+
+
+def build_deterministic_policy(input_dict, obs_space, action_space, config):
+    """Contruct deterministic policy tf graph."""
+    if config["use_state_preprocessor"]:
+        state_preprocessor = ModelCatalog.get_model(
+            input_dict, obs_space, action_space, 1, config["model"]
+        )
+        action_out = state_preprocessor.last_layer
+    else:
+        state_preprocessor = None
+        action_out = input_dict[SampleBatch.CUR_OBS]
+
+    activation = config["actor_hidden_activation"]
+    for hidden in config["actor_hiddens"]:
+        action_out = keras.layers.Dense(units=hidden, activation=activation)(action_out)
+    action_out = keras.layers.Dense(units=action_space.shape[0], activation=None)(
+        action_out
+    )
+
+    # Use sigmoid to scale to [0,1], but also double magnitude of input to
+    # emulate behaviour of tanh activation used in DDPG and TD3 papers.
+    sigmoid_out = tf.nn.sigmoid(2 * action_out)
+    # Rescale to actual env policy scale
+    # (shape of sigmoid_out is [batch_size, dim_actions], so we reshape to
+    # get same dims)
+    action_range = (action_space.high - action_space.low)[None]
+    low_action = action_space.low[None]
+    actions = action_range * sigmoid_out + low_action
```

Comment:
It should be noted that using the sigmoid function this way to force the action values to remain between the bounds of the `action_space` has an multiplicative effect on the gradient of the activation function. In principle, this can be troublesome depending on the scale of the `action_range`...

One possible solution is to also divide the argument given the the sigmoid function by `action_range`. So the gradient of the modified sigmoid remais the same of the original sigmoid, even though the range is now given by `action_range`.